midend/hsIndexSimplify.cpp

```diff
@@ -0,0 +1,127 @@
+#include "midend/hsIndexSimplify.h"
+
+#include <iostream>
+#include <sstream>
+
+namespace P4 {
+
+const IR::Node* HSIndexFinder::postorder(IR::ArrayIndex* curArrayIndex) {
+    if (curArrayIndex->right->is<IR::Constant>()) {
+        return exprIndex2Member(curArrayIndex->type, curArrayIndex->left,
+                                curArrayIndex->right->checkedTo<IR::Constant>());
+    }
+    // Finding the first occurrence of non-concrete array index.
+    if (isFinder) {
+        if (arrayIndex == nullptr && !curArrayIndex->right->is<IR::Constant>()) {
+            arrayIndex = curArrayIndex;
+        }
+    } else {
+        // Translating current array index.
+        if (arrayIndex != nullptr && curArrayIndex->equiv(*arrayIndex)) {
+            return exprIndex2Member(curArrayIndex->type, curArrayIndex->left,
+                                    new IR::Constant(index));
+        }
+    }
+    return curArrayIndex;
+}
+
+size_t HSIndexFinder::getArraySize() {
+    const auto* typeStack = arrayIndex->left->type->checkedTo<IR::Type_Stack>();
+    return typeStack->getSize();
+}
+
+const IR::ArrayIndex* HSIndexFinder::exprIndex2Member(const IR::Type* type,
+                                                      const IR::Expression* expression,
+                                                      const IR::Constant* constant) {
+    return new IR::ArrayIndex(type, expression, constant);
+}
+
+IR::Node* HSIndexSimplifier::eliminateArrayIndexes(IR::Statement* statement) {
+    // Checking non-concrete array indexes.
+    HSIndexFinder aiFinder;
+    const IR::Statement* elseBody = nullptr;
+    bool isIf = false;
```

Comment:
Rewrite this to const auto* isIf = statement->is<IR::IfStatement>() and use that to check.